Referencing C++/WinRT static library results in MSB3272

Originally filed as https://developercommunity.visualstudio.com/content/problem/629524/static-library-reference-causes-there-was-a-proble.html

Repro via @DHowett-MSFT: [DoStaticCppWinrtLibrariesWorkOrNot.zip](https://github.com/microsoft/msbuild/files/3432120/DoStaticCppWinrtLibrariesWorkOrNot.zip) who also developed a workaround: https://github.com/microsoft/terminal/pull/2007/commits/40beb06dde29a39549af7af0dd19ab3dc697a7e1.

```
error MSB3272: There was a problem reading the implementation file "S:\repro\devdiv\DevDiv\948240\ARM\Debug\ComponentLib\ComponentLib.lib". "Unable to read beyond the end of the stream." [S:\repro\devdiv\DevDiv\948240\ComponentDll\ComponentDll.vcxproj]
```

The full exception stack is:

```
   at System.IO.BinaryReader.FillBuffer(Int32 numBytes) in f:\dd\ndp\clr\src\BCL\system\io\binaryreader.cs:line 580
   at System.IO.BinaryReader.ReadUInt32() in f:\dd\ndp\clr\src\BCL\system\io\binaryreader.cs:line 191
   at Microsoft.Build.Tasks.ReferenceTable.ReadMachineTypeFromPEHeader(String dllPath) in /_/src/Tasks/AssemblyDependency/ReferenceTable.cs:line 2905
   at Microsoft.Build.Tasks.ReferenceTable.VerifyArchitectureOfImplementationDll(String dllPath, String winmdFile) in /_/src/Tasks/AssemblyDependency/ReferenceTable.cs:line 2796
```

This is a regression resulting from #4418.

--- a/src/Tasks/AssemblyDependency/ReferenceTable.cs
+++ b/src/Tasks/AssemblyDependency/ReferenceTable.cs
@@ -898,7 +898,9 @@ namespace Microsoft.Build.Tasks
                 {
                     var implementationFile = reference.PrimarySourceItem.GetMetadata(ItemMetadataNames.winmdImplmentationFile);
 
-                    if (!String.IsNullOrEmpty(implementationFile))
+                    // Static library projects can produce a .winmd with an associated .lib, but that is not
+                    // a real ImplementationAssembly--it would fail downstream when trying to read its PE header.
+                    if (!String.IsNullOrEmpty(implementationFile) && Path.GetExtension(implementationFile) == ".dll")
                     {
                         companionFile = Path.Combine(Path.GetDirectoryName(baseName), implementationFile);
                     }

